### DISCUSSION TITLE
Perception warning with Custom Bindings

### DISCUSSION THREAD

**Discussion by wes-nz** [pointfreeco/swift-perception#32] [Q&A] [answered]
Hi,

I have been unable to remove the perception warnings logs when using custom Bindings, are there any ways around it? 

```Swift

public struct MainView: View {
    
    @Perception.Bindable var store: StoreOf<Main>
    
    public var body: some View {
        WithPerceptionTracking {
            Text("hello")
                .sheet(isPresented: Binding(
                    get: {
                        store.isSheetOpen // ⚠️ [Perception] Perceptible state was accessed but is not being tracked.
                    },
                    set: { value in
                        store.send(.sheetUpdated(isPresented: value))
                    }
                )) {
                    Text("world")
                }
        }
    }
}
```
Sometimes using `$store.property` is less convenient than writing my own bindings, even if it is not the preferred way.

Thanks!

<details>

<summary>👉Complete Running Example👈</summary>

```Swift

import SwiftUI
import ComposableArchitecture

@main
struct HelloWorldApp: App {
        
    var body: some Scene {
        WindowGroup {
            MainView(store: StoreOf<Main>(initialState: .init(), reducer: { Main() }))
        }
    }
}


@Reducer
public struct Main {
    
    @ObservableState
    public struct State: Equatable {
        public var isSheetOpen: Bool = false
    }
    
    public enum Action {
        case sheetUpdated(isPresented: Bool)
    }
    
    public var body: some Reducer<State, Action> {
        Reduce { state, action in
            switch action {
            case let .sheetUpdated(isPresented):
                state.isSheetOpen = isPresented
                return .none
            }
        }
    }
    
}

public struct MainView: View {
    @Perception.Bindable var store: StoreOf<Main>
    
    public var body: some View {
        WithPerceptionTracking {
            Button("hello", action: {
                store.send(.sheetUpdated(isPresented: true))
            })
            .sheet(isPresented: Binding(
                get: {
                    store.isSheetOpen // ⚠️ [Perception] Perceptible state was accessed but is not being tracked.
                },
                set: { value in
                    store.send(.sheetUpdated(isPresented: value))
                }
            )) {
                Text("world")
            }
        }
    }
}

```
</details>

**Comment by mbrandonw** [comment, marked as answer]
Hi @wes-nz, unfortunately there's nothing we can do (as far as we know) to detect when bindings are accessed in view bodies vs out of view bodies. Since TCA can control the creation of bindings we perform extra work in them to disable perception checks.

If you really do want to create your own ad hoc bindings, then you can wrap state access inside the `get` with [_withoutPerceptionChecking](https://github.com/pointfreeco/swift-perception/blob/29b059325ae6be84250008b3833b90b13a8e7a3c/Sources/Perception/PerceptionRegistrar.swift#L325) to disable perception checking for a lexical scope.

However, I would highly recommend against creating ad hoc bindings like this. Not only does it run afoul of things like perception checking, but it can also breaking SwiftUI animations in subtle ways. For whatever reason, deriving bindings via dynamic member lookup tends to play better with SwiftUI's tools.

In your case you could simply do `$store.isSheetOpen.sending(\.sheetUpdated)` instead of creating the binding from scratch. Does that not work for you?

**Reply by wes-nz** [reply]
Thank you for the detailed answer @mbrandonw. I wasn't aware of the potential animation issues related to ad hoc bindings. I will do the necessary changes to use `.sending` 👍

**Reply by zhuscat** [reply]
@mbrandonw Hi. Are there any examples of ad hoc binding causing breaking SwiftUI animations?
